### ISSUE TITLE
Rebalance failed due to clojure.lang.ArityException

### ISSUE THREAD

**Issue by Orhideous** [twosigma/Cook#90]
Hello.
When sending several tasks from two different users when the requirements for the CPU for each task equal to the number of all  CPUs in the cluster the following error occurs:

```
2015-11-25 09:45:08,089 ERROR cook.mesos.rebalancer [async-thread-macro-5] - Rebalance failed
clojure.lang.ArityException: Wrong number of args (0) passed to: dru/accumulate-resources/fn--9310
    at clojure.lang.AFn.throwArity(AFn.java:429)
    at clojure.lang.AFn.invoke(AFn.java:28)
    at clojure.core$reductions$fn__6382.invoke(core.clj:6636)
    at clojure.lang.LazySeq.sval(LazySeq.java:40)
    at clojure.lang.LazySeq.seq(LazySeq.java:49)
    at clojure.lang.RT.seq(RT.java:484)
    at clojure.core$seq.invoke(core.clj:133)
    at clojure.core$map$fn__4245.invoke(core.clj:2551)
    at clojure.lang.LazySeq.sval(LazySeq.java:40)
    at clojure.lang.LazySeq.seq(LazySeq.java:49)
    at clojure.lang.RT.seq(RT.java:484)
    at clojure.core$seq.invoke(core.clj:133)
    at clojure.core$map$fn__4252.invoke(core.clj:2568)
    at clojure.lang.LazySeq.sval(LazySeq.java:40)
    at clojure.lang.LazySeq.seq(LazySeq.java:49)
    at clojure.lang.RT.seq(RT.java:484)
    at clojure.core$seq.invoke(core.clj:133)
    at clojure.core.protocols$seq_reduce.invoke(protocols.clj:30)
    at clojure.core.protocols$fn__6078.invoke(protocols.clj:54)
    at clojure.core.protocols$fn__6031$G__6026__6044.invoke(protocols.clj:13)
    at clojure.core$reduce.invoke(core.clj:6289)
    at clojure.core$into.invoke(core.clj:6342)
    at cook.mesos.dru$next_task__GT_scored_task.invoke(dru.clj:95)
    at cook.mesos.rebalancer$next_state.invoke(rebalancer.clj:203)
    at cook.mesos.rebalancer$compute_next_state_and_preemption_decision.invoke(rebalancer.clj:264)
    at cook.mesos.rebalancer$rebalance.invoke(rebalancer.clj:278)
    at cook.mesos.rebalancer$rebalance_BANG_.invoke(rebalancer.clj:298)
    at cook.mesos.rebalancer$start_rebalancer_BANG_$fn__10998.invoke(rebalancer.clj:371)
    at chime$chime_at$fn__9809$state_machine__4452__auto____9810$fn__9812$fn__9818.invoke(chime.clj:68)
    at clojure.core.async$thread_call$fn__4631.invoke(async.clj:405)
    at clojure.lang.AFn.run(AFn.java:22)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at java.lang.Thread.run(Thread.java:745)
```

Entire log available **[on pastebin](http://pastebin.com/0iNDhWLv)**

**Comment by dgrnbrg** [comment]
To clarify, each of these tasks uses as many CPUs as all the machines in the cluster? Or each task uses more the total number of CPUs in the cluster, which would make it unschedulable?

**Comment by Orhideous** [comment]
@dgrnbrg, first case — each of these tasks uses as many CPUs as all the machines in the cluster.

**Comment by dgrnbrg** [comment]
@Orhideous I'm not sure I understand--can you clarify?

Example 1) We have a cluster with 5 machines, each with 4 CPUs and 8 GB RAM. Each user is submitting 100 tasks that use 4 CPUs and 4 GB RAM.

Example 2) We have a cluster with 5 machines, each with 4 CPUs and 8 GB RAM. Each user is submitting 10 tasks that use 20 CPUs and 20 GB RAM.

Example 1 should schedule just fine, example 2 can't run (b/c we'll never find a machine with 20 CPUs).

Do you see tasks running and completing?

**Comment by Orhideous** [comment]
@dgrnbrg I got this error in dev environment on a single machine, trying to run tasks in two Spark shells at the same time, and each of shells used all available 4 CPU.

**Comment by dgrnbrg** [comment]
So you have 1 Mesos agent w/ 4 CPUs and 2 tasks, each using 4 CPUs? I'm trying to wrap my head around the cluster's configuration to trace why and how this code path was encountered, and what the solution is :)

**Comment by Orhideous** [comment]
@dgrnbrg Yes, exactly.

**Comment by dgrnbrg** [comment]
Were you able to successfully run spark tasks from either repl?

**Comment by Orhideous** [comment]
I've reproduced this problem by starting a console like this:
```bash
./spark-shell --executor-memory 2G --total-executor-cores 4 --master cook://first_user:pass_1@localhost:12321
```

Tasks submitted and executed successfully in it, but even if you do nothing, and do not run in parallel more consoles, but just wait, every attempt to rebalance cluster (after hardcoded timeout in Cook) leads to this error.

**Comment by dgrnbrg** [comment]
I think that's a spurious log message when the rebalancer runs and doesn't
do anything. We'll improve the logging here so that it doesn't misleadingly
make you think things are broken :)

**Comment by Orhideous** [comment]
@dgrnbrg even if so, why after the launch of the first interactive pyspark session
`./pyspark --executor-memory 2G --total-executor-cores 4 --master cook://first_user:pass_1@localhost:12321`
with a simple task (`sc.parallelize([1,2,3]).count()`, for example) any task running in the second session
`./pyspark --executor-memory 2G --total-executor-cores 4 --master cook://second_user:pass_1@localhost:12321`
are not performed, and results in an error:
`TaskSchedulerImpl: 71 - Initial job has not accepted any resources; check your cluster UI to ensure that workers are registered and have sufficient resources`

**Comment by dgrnbrg** [comment]
The 2 shells are racing to claim the entire agent. The first shell is able to launch its worker, but the 2nd cannot do so, since there are no resources available.

There's no preemption solution here, since preempting the first shell's worker would harm it as much as it would help the second shell. A more interesting case would be if you had 10 identical agents, and each pyspark shell was launched on that cluster. Initially, one shell should get all the resources, but after rebalancing, they should each have ~1/2.

**Comment by DaoWen** [comment]
I'm going to close this as I think at least the original `ArityException` error is no longer an issue. If the spark-related problems discussed later in this thread are still outstanding, then we can open a new issue.
